File: frontend-v2/src/lib/api/deepResearch.ts
import { api } from './client';

export type DeepResearchStatus =
  | 'queued'
  | 'planning'
  | 'searching'
  | 'reading'
  | 'synthesizing'
  | 'verifying'
  | 'running'
  | 'paused'
  | 'completed'
  | 'failed'
  | 'cancel_requested'
  | 'cancelled';

export type DeepResearchFollowUpMode = 'ask' | 'research';

export interface CitedSource {
  title: string;
  url?: string | null;
  source?: string | null;
  external_id?: string | null;
  /** "academic" (arXiv/Semantic Scholar/Scholar) | "web" (OpenAlex). */
  type?: 'academic' | 'web' | string | null;
  authors?: string[] | string | null;
  year?: number | null;
  citation_count?: number | null;
  pdf_url?: string | null;
}

export interface DeepResearchSession {
  id: number;
  question: string;
  title?: string | null;
  status: DeepResearchStatus;
  last_error_code?: string | null;
  created_at: string;
  updated_at: string;
  current_generation?: number;
  lifecycle_version?: number;
}

export interface DeepResearchSessionDetail extends DeepResearchSession {
  report?: string | null;
  cited_sources?: CitedSource[] | null;
  generation?: DeepResearchGenerationSummary | null;
}

export type DeepResearchVerificationStatus =
  | 'pending'
  | 'in_progress'
  | 'verified'
  | 'insufficient_evidence'
  | 'needs_attention';

export interface DeepResearchGenerationSummary {
  id: number;
  generation_number: number;
  mode: DeepResearchFollowUpMode;
  status: DeepResearchStatus;
  provider_type?: string | null;
  model?: string | null;
  scope: string;
  effort: string;
  phase: string;
  progress: number;
  source_count: number;
  verification_status: DeepResearchVerificationStatus;
  stop_reason?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
}

export interface DeepResearchArchiveResponse {
  items: DeepResearchSession[];
  total: number;
  limit: number;
  offset: number;
  has_more: boolean;
}

export interface DeepResearchMessage {
  id: number;
  session_id: number;
  generation_id: number;
  generation_number: number;
  role: 'user' | 'assistant';
  mode: DeepResearchFollowUpMode;
  content: string;
  source_ids: number[];
  verification?: string | null;
  created_at: string;
}

export interface DeepResearchFollowUpResponse {
  mode: DeepResearchFollowUpMode;
  status: DeepResearchStatus;
  generation_number: number;
  message: DeepResearchMessage;
  assistant_message?: DeepResearchMessage | null;
  session: DeepResearchSession;
}

function createIdempotencyKey(): string {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return crypto.randomUUID();
  }
  return `dr-${Date.now()}-${Math.random().toString(36).slice(2)}`;
}

export const deepResearchApi = {
  start: (
    question: string,
    idempotencyKey = createIdempotencyKey(),
  ): Promise<DeepResearchSession> =>
    api.post<DeepResearchSession>(
      '/deep-research',
      { question },
      { headers: { 'Idempotency-Key': idempotencyKey } },
    ),

  list: (limit = 50, offset = 0): Promise<DeepResearchSession[]> =>
    api.get<DeepResearchSession[]>('/deep-research', { params: { limit, offset } }),

  archive: (q = '', limit = 20, offset = 0): Promise<DeepResearchArchiveResponse> =>
    api.get<DeepResearchArchiveResponse>('/deep-research/archive', {
      params: { q: q.trim() || undefined, limit, offset },
    }),

  get: (sessionId: number): Promise<DeepResearchSessionDetail> =>
    api.get<DeepResearchSessionDetail>(`/deep-research/${sessionId}`),

  messages: (sessionId: number): Promise<DeepResearchMessage[]> =>
    api.get<DeepResearchMessage[]>(`/deep-research/${sessionId}/messages`),

  followUp: (
    sessionId: number,
    mode: DeepResearchFollowUpMode,
    question: string,
    idempotencyKey = createIdempotencyKey(),
  ): Promise<DeepResearchFollowUpResponse> =>
    api.post<DeepResearchFollowUpResponse>(
      `/deep-research/${sessionId}/messages`,
      { mode, question },
      { headers: { 'Idempotency-Key': idempotencyKey } },
    ),

  remove: (sessionId: number): Promise<{ message: string; id: number }> =>
    api.delete<{ message: string; id: number }>(`/deep-research/${sessionId}`),

  cancel: (sessionId: number): Promise<DeepResearchSession> =>
    api.post<DeepResearchSession>(`/deep-research/${sessionId}/cancel`),

  resume: (sessionId: number): Promise<DeepResearchSession> =>
    api.post<DeepResearchSession>(`/deep-research/${sessionId}/resume`),
};

export default deepResearchApi;

File: backend/app/schemas/deep_research.py
from datetime import datetime
from typing import Any, List, Literal, Optional

from pydantic import BaseModel, Field

from app.core.config import settings


class DeepResearchSessionCreate(BaseModel):
  """Request to start a deep-research run."""

  question: str = Field(min_length=1, max_length=settings.DEEP_RESEARCH_MAX_QUESTION_LENGTH)
  provider_id: Optional[int] = Field(default=None, gt=0)


class CitedSource(BaseModel):
  """A source surfaced during a research run.

  Rich metadata is populated when the source came from a discovery tool
  (structured paper); report-link fallbacks carry only title + url.
  """

  title: str
  url: Optional[str] = None
  source: Optional[str] = None
  external_id: Optional[str] = None
  # "academic" (arXiv/Semantic Scholar/Scholar) | "web" (OpenAlex)
  type: Optional[str] = None
  authors: Optional[Any] = None
  year: Optional[int] = None
  citation_count: Optional[int] = None
  pdf_url: Optional[str] = None


class DeepResearchSession(BaseModel):
  """Deep-research session summary (list view)."""

  id: int
  question: str
  title: Optional[str] = None
  status: str
  last_error_code: Optional[str] = None
  current_generation: int = 1
  lifecycle_version: int = 0
  created_at: datetime
  updated_at: datetime

  class Config:
    from_attributes = True


class DeepResearchGenerationSummary(BaseModel):
  """Safe progress metadata for the current execution generation.

  Provider credentials and agent checkpoints deliberately do not cross this
  boundary. The values are a snapshot, so clients can render recovery and
  progress without treating a dropped stream as a failed run.
  """

  id: int
  generation_number: int
  mode: Literal["ask", "research"]
  status: str
  provider_type: Optional[str] = None
  model: Optional[str] = None
  scope: str
  effort: str
  phase: str
  progress: int = Field(ge=0, le=100)
  source_count: int = Field(default=0, ge=0)
  verification_status: Literal[
    "pending", "in_progress", "verified", "insufficient_evidence", "needs_attention"
  ]
  stop_reason: Optional[str] = None
  started_at: Optional[datetime] = None
  finished_at: Optional[datetime] = None


class DeepResearchArchiveResponse(BaseModel):
  """Bounded archive page with an explicit continuation contract."""

  items: List[DeepResearchSession]
  total: int = Field(ge=0)
  limit: int = Field(ge=1, le=100)
  offset: int = Field(ge=0)
  has_more: bool


class DeepResearchSessionDetail(DeepResearchSession):
  """Deep-research session with the report and cited sources.

  Internal execution checkpoints are intentionally omitted — they are never
  sent to the client.
  """

  report: Optional[str] = None
  cited_sources: Optional[List[CitedSource]] = None
  generation: Optional[DeepResearchGenerationSummary] = None


class DeepResearchFollowUpCreate(BaseModel):
  """Explicit conversational mode for a completed research session."""

  mode: Literal["ask", "research"]
  question: str = Field(min_length=1, max_length=settings.DEEP_RESEARCH_MAX_QUESTION_LENGTH)


class DeepResearchMessage(BaseModel):
  """Safe durable conversation turn; internal payloads are never exposed."""

  id: int
  session_id: int
  generation_id: int
  generation_number: int
  role: Literal["user", "assistant"]
  mode: Literal["ask", "research"]
  content: str
  source_ids: List[int] = Field(default_factory=list)
  verification: Optional[str] = None
  created_at: datetime


class DeepResearchFollowUpResponse(BaseModel):
  """Result of an Ask this research or Research further request."""

  mode: Literal["ask", "research"]
  status: str
  generation_number: int
  message: DeepResearchMessage
  assistant_message: Optional[DeepResearchMessage] = None
  session: DeepResearchSession
